Core/include/Acts/Propagator/GuidedNavigator.hpp

```diff
@@ -0,0 +1,230 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2021 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/Algebra.hpp"
+#include "Acts/Geometry/GeometryIdentifier.hpp"
+#include "Acts/Geometry/TrackingGeometry.hpp"
+#include "Acts/Geometry/TrackingVolume.hpp"
+#include "Acts/Surfaces/Surface.hpp"
+#include "Acts/Utilities/Logger.hpp"
+
+namespace Acts {
+
+template <typename Guider>
+class GuidedNavigator {
+  Guider m_guider;
+  std::shared_ptr<const TrackingGeometry> m_tgeo;
+
+ public:
+  struct State {
+    std::vector<SurfaceIntersection> navSurfaces;
+    std::vector<SurfaceIntersection>::const_iterator navSurfaceIter;
+    const Surface *currentSurface;
+    const Surface *startSurface;
+
+    bool navigationBreak = false;
+    bool targetReached = false;
+
+    const TrackingVolume *currentVolume = nullptr;
+    const TrackingVolume *worldVolume = nullptr;
+    const Surface *targetSurface = nullptr;
+  };
+
+  // TODO Must the Navigator really be default constructable?
+  // See Propagator.hpp:247
+  GuidedNavigator() { throw_assert(false, "Don't use this constructor!"); }
+
+  /// Constructor of the ML Navigator
+  GuidedNavigator(Guider &&guider,
+                  std::shared_ptr<const Acts::TrackingGeometry> tgeo)
+      : m_guider(guider), m_tgeo(tgeo) {}
+
+  /// The whole purpose of this function is to set currentSurface, if possible
+  template <typename propagator_state_t, typename stepper_t>
+  void status(propagator_state_t &state, const stepper_t &stepper) const {
+    const auto &logger = state.options.logger;
+    ACTS_VERBOSE(">>>>>>>> STATUS <<<<<<<<<");
+
+    if (state.navigation.navigationBreak)
+      return;
+
+    // TODO Is initialization correct for all cases?
+    if (state.navigation.navSurfaces.empty()) {
+      initialize(state, stepper);
+      return;
+    }
+
+    // Navigator status always resets the current surface
+    state.navigation.currentSurface = nullptr;
+
+    // Establish the surface status
+    auto surfaceStatus = stepper.updateSurfaceStatus(
+        state.stepping, *state.navigation.navSurfaceIter->representation,
+        false);
```

Comment:
Can't `navSurfaceIter` be `.end()` here?

Reply:
At least from what I see not, since the `.target(...)` call does set the `navigationBreak`-flag, which causes an early return in the `status(...)` call.
The only possibility seems to me an empty surface-vector, which is checked by assertions (which are not compiled in the default build mode^^)

But yeah, one could also do an additional check and set the `navigationBreak`-flag accordingly...